File: 17-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.py
from typing import List

class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if not digits:
            return []
        
        mapping = {
            "2": "abc", "3": "def", "4": "ghi", "5": "jkl",
            "6": "mno", "7": "pqrs", "8": "tuv", "9": "wxyz"
        }
        
        res = []

        def backtrack(index: int, path: str):
            if index == len(digits):
                res.append(path)
                return
            for c in mapping[digits[index]]:
                backtrack(index + 1, path + c)

        backtrack(0, "")
        return res

File: 17-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<string> letterCombinations(string digits) {
        if (digits.empty()) return {};
        vector<string> mapping = {
            "", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"
        };
        vector<string> res;
        string curr;

        function<void(int)> backtrack = [&](int idx) {
            if (idx == digits.size()) {
                res.push_back(curr);
                return;
            }
            for (char c : mapping[digits[idx] - '0']) {
                curr.push_back(c);
                backtrack(idx + 1);
                curr.pop_back();
            }
        };

        backtrack(0);
        return res;
    }
};
